### PR TITLE
Open62541 compat fixes for iseg

### PR DESCRIPTION
**Pull request by ben-farnham**
Hi All,

Changes made for an open62541 build of the ISEG OPC-UA server (now builds with either Unified Automation or open62541 - quite nice!).

As usual I only provided interface/implementation on as as-needs basis, just enough for the ISEG server to compile and link; basically small enhancements to the UaDateTime and UaVariant classes.

For review...

Cheers,
Ben

### CODE REVIEW

**Comment by pnikiel** [comment]
Hi Ben,
I've got few questions:
1. bool isBad() const { return UA_StatusCode(m_status) >= OpcUa_Bad; }

IMO this will not work unless there is some magic within sign casting, i.e. OpcUa_Good = 0, and OpcUa_Bad is  -2^31 so zero will be always greater than something which is negative. therefore isBad(Good)= true which makes no sense.
If the point is to (invisibly) case m_status to unsigned and this way check for 31st bit to be on, I'd prefer to do it using bitwise operator (which works the same way on both signed and unsigned), as less ambiguous an more portable.
1. UaVariant::operator==
   is there a function that does it in open62541? 
   I'd prefer to avoid handling internals of open62541 data representation because there should be functions ready to do it, plus it is not future proof and (IMO) we have too little knowledge of such intricate details.
2. UaDateTime::initializeInternalDateTimeStruct(UA_DateTimeStruct& dateTimeStruct)
   again, do we have functions for this in open62541?
3. void UaDateTime::addMilliSecs(int msecs)
   {
   m_dateTime.milliSec += msecs;
   }

IMO overrun of milliseconds counter not handled. Probably the same for addSecs, depending on what is the "master" time unit.

5.
Time formatting by boost.
Boost is generally not used in the compat module; the only exception being FOREACH macro which can be easily rewritten and doesnt need any external dependencies. 
My point: I'd like to first discuss with other developers if they want to do this trade: start being dependent on boost to bring time formatting feats.
Are there time formatting functions from open62541 itself?

**Comment by ben-farnham** [comment]
Hi Piotr,
Thanks for a thorough review - appreciated.

I'll look again to the implementations for 
- bool isBad()
- void UaDateTime::addMilliSecs()

Quite possibly I've missed something there - I'll check, more to follow in another post...

As regards the comments on
- UaVariant::operator==
- UaDateTime::initializeInternalDateTimeStruct

No, not that I saw anyway - documentation for [UaVariant](http://open62541.org/doc/current/types.html#variant) and [DateTime](http://open62541.org/doc/current/types.html#datetime) doesn't mention comparison/initialisation methods, nor did I find anything trawling around in the header files. I don't think open62541 provides that level of 'niceness', I assume clean initialisation and variant comparison is the responsibility of the consumer environment - in our case the open62541-compat code. If you've seen something otherwise though I'm happy to use it.
Our long term goal should be to have compat layer unit tests (e.g. googletests) to ensure the compat layer behaves predictably, for, say, variant comparison, so when we upgrade to a newer open62541 version (perhaps with subtle changes to the variant structure) our tests pass: so we're OK, or they fail, so we have to inspect the compat layer implementation.
- Time formatting by boost.

The perennial question: to boost or not to boost... Well, no point in going to nth level detail here (I hadn't realised we were avoiding boost in the compat layer) but the usual story applies: it works, it's cross platform and (with careful application) can reduce complexity a bit . Obviously the UaDateTime toString/fromString implementations can be rewritten without boost - but boost is so very handy. If there's a strong anti-boost sentiment I'll rewrite it, otherwise not.

Cheers,
Ben

**Comment by pnikiel** [comment]
Hi Ben,
I've got a superb idea: for variant comparison and initializing open62541's datetime, in case these things REALLY do not exist in open6, why not to contribute them to open62541? You would have a contribution to a project which matters, we would stand in awe, and they would get fixed in proper place. I still think that in the compat layer we should not touch internal data representation of open62541 as much as possible.

It's not a joke; I'm really suggesting you contribute this to open6.

**Comment by ben-farnham** [comment]
Hmm, 

Disappointing, I was investigating time formatting using std C++ (rather than boost). From what I can see online it appears that the key std methods driving time formatting are
[std::get_time](http://en.cppreference.com/w/cpp/io/manip/get_time)
[std::put_time](http://en.cppreference.com/w/cpp/io/manip/put_time)

Attempting to compile this code
`
#include \<iostream\>
#include \<sstream\>
#include \<locale\>
#include \<iomanip\>

int main()
{
    std::tm t = {};
    std::istringstream ss("1970-01-01T00:00:00Z");
    ss >> std::get_time(&t, "%Y-%m-%dT%H:%M:%S%ZP");
    
    if (ss.fail()) 
    {
        std::cout << "Parse failed\n";
    } 
    else
    {   
        std::cout << std::put_time(&t, "%c") << '\n';
    }
}
`

command line

> g++ main.cpp -o std_time_handling

gives

> main.cpp: In function \u2018int main()\u2019:
main.cpp:10:11: error: \u2018get_time\u2019 is not a member of \u2018std\u2019
     ss >> std::get_time(&t, "%Y-%m-%dT%H:%M:%S%ZP");
           ^
main.cpp:18:22: error: \u2018put_time\u2019 is not a member of \u2018std\u2019
         std::cout << std::put_time(&t, "%c") << '\n';


g++ version info

> bash-4.2$ g++ --version
g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-4)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

So, a question for the community - anyone know of some other part of the C++ standard that provides concise time/string handling. I do not want to write (or maintain) a load of string parsing code, better to use boost in that case.

Cheers,
Ben

**Comment by pnikiel** [comment]
Hi Ben,
could you try compiling this way:
g++ -std=c++0x main.cpp -o std_time_handling 
or gnu++0x instead of c++0x

?

**Comment by ben-farnham** [comment]
bash-4.2$ g++ -std=c++0x main.cpp -o std_time_handling
main.cpp: In function \u2018int main()\u2019:
main.cpp:10:11: error: \u2018get_time\u2019 is not a member of \u2018std\u2019
     ss >> std::get_time(&t, "%Y-%m-%dT%H:%M:%S%ZP");
           ^
main.cpp:18:22: error: \u2018put_time\u2019 is not a member of \u2018std\u2019
         std::cout << std::put_time(&t, "%c") << '\n';
                      ^

**Comment by ben-farnham** [comment]
bash-4.2$ g++ -std=gnu++0x main.cpp -o std_time_handling
main.cpp: In function \u2018int main()\u2019:
main.cpp:10:11: error: \u2018get_time\u2019 is not a member of \u2018std\u2019
     ss >> std::get_time(&t, "%Y-%m-%dT%H:%M:%S%ZP");
           ^
main.cpp:18:22: error: \u2018put_time\u2019 is not a member of \u2018std\u2019
         std::cout << std::put_time(&t, "%c") << '\n';
                      ^

**Comment by ben-farnham** [comment]
Aha,
Apparently [a known bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54354), functionality is available in gcc v5+

My proposed workaround in the meantime: boost. 

Better ideas are welcome, but... this branch contains fixes to build the ISEG OPC-UA server using the open62541 backend. Right now, only Unified Automation license holders can build this server from latest - that blocks mainline collaboration with commercial partners without a license. Dithering about merging master -> branch etc is possible, but maintaining the fight against branch divergence costs time (a scarce resource).

So, the practical, pragmatic reason for being pro-boost; it provides a clean implementation that helps ease up a somewhat blocked collaboration. From my perspective, the no-boost approach benefit is less clear. Furthermore, as it stands today this module is, and always has been, boost dependent (BOOST_FOREACH).

In summary then: I want to merge the boost dependent time/string handler code to master, objections?

Cheers,
Ben

**Comment by pnikiel** [comment]
Hi Ben,

I agree with you - I do not see any other sensible solution. But please make sure you require a sane version like 1.48 which is in CC7 by default.

cheers p

**Comment by ben-farnham** [comment]
Hi All,

A few changes with respect to the review comments:
1. UaDateTime: implementation is now more based on open62541's UA_DateTime (a typedef of a 64 bit integer) rather than open62541's UA_DateTimeStruct. A slightly cleaner implementation in the end as regards functions addSec/addMillisec.
2. Status.isBad(): a bit check rather than > test. Checks if bit 31 is on.

So, let me know what you think of v2.0 - from my side I'm ready to merge.

Cheers,
Ben

**Comment by ben-farnham** [comment]
BTW: variant operator== comparison remains. However, I'll ask open62541.org if they would like an equality test implementation for variants. If they do and it's accepted I will replace the compat layer implementation to a call to open62541's implementation. But that process may take a while.

Cheers,
Ben

**Comment by pnikiel** [comment]
Hi Ben,
for operator== business, could you just open a ticket in our JIRA for this and put JIRA task name with a note comment close to the implementation in the code that this is temporary solution?

I have no comments for other aspects, thanks for your work!

**Comment by ben-farnham** [comment]
Hi All,
For info, here's the link to the related issue I just raised for open62541.
https://github.com/open62541/open62541/issues/902

Let's see what they think. I'll add a JIRA ticket.
Cheers,
Ben
